# Test Score of Students

In [1]:
import pandas as pd
import numpy as np

In [2]:
std_data = pd.read_csv("DS1_C4_S4_Test_Scores_Data_Practice.csv")

In [3]:
std_data.shape

(2133, 10)

## TASK - 1:

### Load the data from the file into a data frame and identify the missing values (if any) using the command, treat the missing values with the most frequently found data in that column. Identify the most freuently found data for any column with missing values.

In [4]:
std_data.isnull().sum()

school                0
school_setting        0
school_type           0
classroom             0
teaching_method       0
student_id            0
gender               10
tuitionfee_waiver     0
test1                 4
test2                 4
dtype: int64

In [5]:
std_data["gender"].value_counts()

gender
Male      1071
Female    1052
Name: count, dtype: int64

In [6]:
std_data["gender"] = std_data["gender"].fillna("Male")
std_data["test1"] = std_data["test1"].fillna(std_data["test1"].mean())
std_data["test2"] = std_data["test2"].fillna(std_data["test2"].median())

In [7]:
std_data.isnull().sum()

school               0
school_setting       0
school_type          0
classroom            0
teaching_method      0
student_id           0
gender               0
tuitionfee_waiver    0
test1                0
test2                0
dtype: int64

## 

## TASK - 2 :

### Identify and display the duplicate rows (if any). Delete the duplicate rows in the data.

In [8]:
std_data.shape

(2133, 10)

In [9]:
std_data[std_data.duplicated(keep="first")].shape

(0, 10)

Interpretation : There is no any duplicate record found in given data.

## 

## TASK - 3 :

### In the column named "tuitionfee_waiver", replace the value "Does not quality" and "Qualifies for reduced/free" with No and Yes, respectively. Display the updated column data and verify if the values are replaced.

In [10]:
std_data = std_data.replace(to_replace=['Does not qualify', 'Qualifies for reduced/free '],
                value=['No', 'Yes'])
std_data

school school_setting school_type classroom teaching_method student_id  \
0     ANKYI          Urban  Non-public       6OL        Standard      2FHT3   
1     ANKYI          Urban  Non-public       6OL        Standard      3JIVH   
2     ANKYI          Urban  Non-public       6OL        Standard      3XOWE   
3     ANKYI          Urban  Non-public       6OL        Standard      556O0   
4     ANKYI          Urban  Non-public       6OL        Standard      74LOE   
...     ...            ...         ...       ...             ...        ...   
2128  ZOWMK          Urban      Public       ZBH        Standard      T8LSK   
2129  ZOWMK          Urban      Public       ZBH        Standard      VNP26   
2130  ZOWMK          Urban      Public       ZBH        Standard      YDR1Z   
2131  ZOWMK          Urban      Public       ZBH        Standard      YUEIH   
2132  ZOWMK          Urban      Public       ZBH        Standard      ZVCQ8   

      gender tuitionfee_waiver  test1  test2  
0     Female                No   62.0   72.0  
1     Female                No   66.0   79.0  
2       Male                No   64.0   76.0  
3     Female                No   61.0   77.0  
4       Male                No   64.0   76.0  
...      ...               ...    ...    ...  
2128  Female                No   39.0   55.0  
2129  Female               Yes   38.0   46.0  
2130  Female               Yes   45.0   51.0  
2131    Male               Yes   46.0   53.0  
2132    Male               Yes   41.0   48.0  

[2133 rows x 10 columns]

## 

## TASK - 4 :

### Calculate and display the average of test1 scores of students at each school. Display the name of the school that has the highest average test1 score.

In [11]:
avg_test_score = std_data.groupby(by=["school"])["test1"].mean()
high_avg_test_score =avg_test_score.sort_values(ascending=False).head(1)
print("The highest average test1 score is of school : ")
high_avg_test_score

The highest average test1 score is of school : 


school
UKPGS    78.453125
Name: test1, dtype: float64

The UKPGS school has highest average test1 score that is 78.45

## 

## TASK - 5 :

### Display the highest test2 score of each school in a tabular format. Display the name of the top five schools based on the output of the above task.

In [12]:
max_score = pd.pivot_table(std_data, index=["school"], values=["test2"], aggfunc = "max")
max_score

test2
school       
ANKYI    79.0
CCAAW    91.0
CIMBB    88.0
CUQAM    76.0
DNQDD    79.0
FBUMG    88.0
GJJHK    85.0
GOKXL    77.0
GOOBU    68.0
IDGFP   100.0
KFZMY    67.0
KZKKE    62.0
LAYPA    84.0
OJOBU    84.0
QOQTS    85.0
UAGPU    83.0
UKPGS    99.0
UUUQX    91.0
VHDHF    77.0
VKWQH    82.0
VVTVA    68.0
ZMNYA    95.0
ZOWMK    63.0

In [13]:
top_five = max_score.sort_values("test2",ascending=False).head()
print("Top five schools are:\n")
top_five

Top five schools are:



test2
school       
IDGFP   100.0
UKPGS    99.0
ZMNYA    95.0
CCAAW    91.0
UUUQX    91.0

## 

## TASK - 6 :

### The school named "FBUMG" is offering a scholarship of 500 for the students of class "5LQ" who have scored >=80 in test2. Display the information of these students along with student_ID and scholarship amounts in a tabular format.

In [14]:
scholarship = std_data[(std_data["school"]=="FBUMG") & (std_data["classroom"]== "5LQ") & (std_data["test2"]>=80)]
scholarship.insert(10,"Scholarship",np.full(15,500))
scholarship 

school school_setting school_type classroom teaching_method student_id  \
467  FBUMG          Rural  Non-public       5LQ    Experimental      04DG5   
468  FBUMG          Rural  Non-public       5LQ    Experimental      20M2D   
469  FBUMG          Rural  Non-public       5LQ    Experimental      39KCW   
470  FBUMG          Rural  Non-public       5LQ    Experimental      5Z1B6   
471  FBUMG          Rural  Non-public       5LQ    Experimental      6TLU8   
472  FBUMG          Rural  Non-public       5LQ    Experimental      ATQQJ   
474  FBUMG          Rural  Non-public       5LQ    Experimental      B9FSU   
475  FBUMG          Rural  Non-public       5LQ    Experimental      EV13K   
476  FBUMG          Rural  Non-public       5LQ    Experimental      I5H37   
477  FBUMG          Rural  Non-public       5LQ    Experimental      JC5I9   
478  FBUMG          Rural  Non-public       5LQ    Experimental      JPE2J   
479  FBUMG          Rural  Non-public       5LQ    Experimental      MFBYU   
480  FBUMG          Rural  Non-public       5LQ    Experimental      O144X   
481  FBUMG          Rural  Non-public       5LQ    Experimental      OGKP3   
484  FBUMG          Rural  Non-public       5LQ    Experimental      YRN9S   

     gender tuitionfee_waiver  test1  test2  Scholarship  
467  Female               Yes   63.0   84.0          500  
468  Female               Yes   68.0   85.0          500  
469    Male               Yes   66.0   81.0          500  
470    Male                No   68.0   88.0          500  
471  Female                No   66.0   81.0          500  
472  Female                No   68.0   82.0          500  
474    Male                No   69.0   83.0          500  
475  Female                No   66.0   83.0          500  
476    Male               Yes   65.0   83.0          500  
477    Male                No   58.0   80.0          500  
478    Male                No   59.0   81.0          500  
479  Female                No   73.0   82.0          500  
480    Male                No   67.0   86.0          500  
481    Male                No   72.0   83.0          500  
484    Male                No   68.0   87.0          500

## 

## TASK - 7 :

### Display the count of each type of teaching method, segregated by school setting.

In [16]:
pd.crosstab(std_data.school_setting, std_data.teaching_method, rownames=["school_setting"], colnames=["teaching_method"])

teaching_method  Experimental  Standard
school_setting                         
Rural                     201       309
Suburban                  284       433
Urban                     275       631

## 

## TASK - 8 :

### Display the gender-wise count of students in different school settings - rural, subrural and urban.

In [17]:
std_data.groupby(by=["school_setting","gender"])["gender"].count()

school_setting  gender
Rural           Female    228
                Male      282
Suburban        Female    368
                Male      349
Urban           Female    456
                Male      450
Name: gender, dtype: int64

## 

## TASK - 9 :

### Display the average test2 scores of male and female students from each school setting.

In [18]:
std_data.groupby(by=["school_setting","gender"])["test2"].mean()

school_setting  gender
Rural           Female    63.991228
                Male      64.092199
Suburban        Female    76.027174
                Male      75.991404
Urban           Female    61.179825
                Male      62.424444
Name: test2, dtype: float64

## 